Source/WebCore/platform/graphics/SourceBufferPrivate.cpp

```diff
@@ -614,27 +613,111 @@ bool SourceBufferPrivate::validateInitializationSegment(const SourceBufferPrivat
     return true;
 }
 
-void SourceBufferPrivate::didReceiveSampleForTrackId(uint64_t, Ref<MediaSample>&& mediaSample)
+void SourceBufferPrivate::didReceiveSample(Ref<MediaSample>&& sample)
+{
+    if (!isAttached())
+        return;
+
+    if (m_pendingAppendOperations.isEmpty())
+        m_pendingAppendOperations.append({ { }, { } });
+
+    DEBUG_LOG(LOGIDENTIFIER, sample.get());
+    m_pendingAppendOperations.last().mediaSamples.append(WTFMove(sample));
+}
+
+void SourceBufferPrivate::append(Ref<SharedBuffer>&& buffer)
 {
-    didReceiveSample(WTFMove(mediaSample));
+    append(buffer->extractData());
 }
 
-void SourceBufferPrivate::didReceiveSample(Ref<MediaSample>&& originalSample)
+void SourceBufferPrivate::append(Vector<unsigned char>&&)
 {
-    if (!m_isAttached || m_didReceiveSampleErrored) {
-        m_pendingMediaSamples.clear();
-        return;
+    RELEASE_ASSERT_NOT_REACHED();
+}
+
+void SourceBufferPrivate::processPendingOperations()
+{
+    while (!m_pendingAppendOperations.isEmpty()) {
+        auto& operation = m_pendingAppendOperations.first();
+        if (operation.initOperation) {
+            processInitSegment();
+            if (m_didReceiveInitializationSegmentErrored)
+                break;
+            if (operation.initOperation)
+                return; // will continue once Init completionHandler is run.
+        }
+        processMediaSamples();
+        if (m_didReceiveSampleErrored)
+            break;
+        m_pendingAppendOperations.removeFirst();
+    };
+
+    auto [parsingSucceeded, isEnded, preTask] = WTFMove(*m_appendCompletedPending);
+    m_appendCompletedPending.reset();
+    parsingSucceeded &= !m_didReceiveInitializationSegmentErrored && !m_didReceiveSampleErrored;
+    appendCompleted(parsingSucceeded, isEnded, WTFMove(preTask));
+}
+
+void SourceBufferPrivate::abortPendingOperations()
+{
+    for (auto& operation : m_pendingAppendOperations) {
+        if (operation.initOperation)
+            operation.initOperation->completionHandler(ReceiveResult::AppendError);
     }
+    m_pendingAppendOperations.clear();
+}
+
+void SourceBufferPrivate::processInitSegment()
+{
+    ASSERT(!m_pendingAppendOperations.isEmpty());
 
-    if (processingInitializationSegment()) {
-        DEBUG_LOG(LOGIDENTIFIER, originalSample.get());
-        // TODO: MediaSample should use uint64_t for trackID.
-        auto trackId = parseInteger<uint64_t>(originalSample->trackID());
-        ASSERT(trackId);
-        m_pendingMediaSamples.append(std::make_pair(*trackId, WTFMove(originalSample)));
+    auto& initOperation = *m_pendingAppendOperations.first().initOperation;
```

Comment:
Should processInitSegment take the operation as a parameter?

Reply:
I could, and move the check logic from processInitSegment to processPendingOperation instead. However, it would need to change the logic in the completionHandler where it's now up to processPendingOperation to remove the initOperation.

I'm not sure if that would be a positive gain there.